modules/rostests/apitests/shell32/closewnd.cpp

```diff
@@ -61,30 +88,31 @@ HWND FindNewWindow(PWINDOW_LIST List1, PWINDOW_LIST List2)
     return NULL;
 }
 
-#define TRIALS_COUNT 8
+static void WaitForForegroundWindow(HWND hWnd, UINT wait = 250)
+{
+    for (UINT waited = 0, interval = 50; waited < wait; waited += interval)
+    {
+        if (GetForegroundWindow() == hWnd || !IsWindowVisible(hWnd))
+            return;
+        Sleep(interval);
+    }
+}
 
 void CloseNewWindows(PWINDOW_LIST List1, PWINDOW_LIST List2)
 {
-    INT cDiff = List2->m_chWnds - List1->m_chWnds;
-    for (INT j = 0; j < cDiff; ++j)
+    for (SIZE_T i = 0; i < List2->m_chWnds; ++i)
     {
-        HWND hWnd = FindNewWindow(List1, List2);
-        if (!hWnd)
-            break;
-
-        for (INT i = 0; i < TRIALS_COUNT; ++i)
-        {
-            if (!IsWindow(hWnd))
-                break;
+        HWND hWnd = List2->m_phWnds[i];
+        if (!IsWindow(hWnd) || FindInWindowList(*List1, hWnd))
+            continue;
 
-            SwitchToThisWindow(hWnd, TRUE);
+        SwitchToThisWindow(hWnd, TRUE);
+        WaitForForegroundWindow(hWnd);
 
-            // Alt+F4
```

Comment:
This Alt-F4 simulation was indeed hackish.

Reply:
It is still not perfect, something deep inside the ROS window manager sometimes chokes when there are this many windows open. I can see it happen when I run the test many times with the same instance of Explorer.exe running the entire time. After a while, you get open Explorer windows that you cannot close and they don't paint correctly.

Reply:
Can we use `TerminateProcess ` at the end to make sure they are all killed?

Reply:
Some of the other tests that use SEE_MASK_NOCLOSEPROCESS can kill it's child process. Some actions (open folder in Explorer) doesn't normally create a child process with a window.

If we call GetWindowThreadProcessId+OpenProcess on the windows we want to close, we will end up killing the main taskbar shell process if we don't add code to filter that out.

Reply:
Yeah, but eventually we should probably look into that.

Reply:
We'll see how much the testbot of improves. In my VM, after 3-4 runs of the ShellExecuteEx test, things start to go wrong but I feel like those unclosable windows I end up with are signs of deeper bugs and not the fault of this test.